Sources/Benchmark/BenchmarkRunner.swift

```diff
@@ -51,27 +51,92 @@ public struct BenchmarkRunner {
+        }
+
         var measurements: [Double] = []
-        measurements.reserveCapacity(settings.iterations)
+        if let n = settings.iterations {
+            measurements = doNIterations(n, benchmark: benchmark, suite: suite)
+        } else {
+            measurements = doAdaptiveIterations(
+                benchmark: benchmark, suite: suite, settings: settings)
+        }
 
-        if settings.warmupIterations > 0 {
-            for _ in 1...settings.warmupIterations {
-                benchmark.run()
-            }
+        let result = BenchmarkResult(
+            benchmarkName: benchmark.name,
+            suiteName: suite.name,
+            measurements: measurements)
+        results.append(result)
+    }
+
+    /// Heuristic for finding good next number of iterations to try, ported from google/benchmark.
+    func predictNumberOfIterationsNeeded(_ measurements: [Double], settings: BenchmarkSettings)
+        -> Int
+    {
+        let minTime = settings.minTime
+        let iters = measurements.count
+
+        // See how much iterations should be increased by.
+        // Note: Avoid division by zero with max(seconds, 1ns)
+        let timeInSeconds = measurements.reduce(0, +) / 1000000000.0
+        var multiplier: Double = minTime * 1.4 / max(timeInSeconds, 1e-9)
+
+        // If our last run was at least 10% of --min-time then we
+        // use the multiplier directly.
+        // Otherwise we use at most 10 times expansion.
+        // NOTE: When the last run was at least 10% of the min time the max
+        // expansion should be 14x.
+        let isSignificant = (timeInSeconds / minTime) > 0.1
+        multiplier = isSignificant ? multiplier : min(10.0, multiplier)
+        if multiplier < 1.0 {
+            multiplier = 2.0
         }
 
-        for _ in 1...settings.iterations {
+        // So what seems to be the sufficiently-large iteration count? Round up.
+        let maxNextIters: Int = Int(max(multiplier * Double(iters), Double(iters) + 1.0).rounded())
+
+        // But we do have *some* sanity limits though..
+        let nextIters = min(maxNextIters, settings.maxIterations)
+
+        return nextIters
+    }
+
+    /// Heuristic when to stop looking for new number of iterations, ported from google/benchmark.
+    func shouldReportResults(_ measurements: [Double], settings: BenchmarkSettings) -> Bool {
+        let tooManyIterations = measurements.count > settings.maxIterations
+        let timeInSeconds = measurements.reduce(0, +) / 1000000000.0
+        let timeIsLargeEnough = timeInSeconds > settings.minTime
+        return tooManyIterations || timeIsLargeEnough
+    }
+
+    func doAdaptiveIterations(
+        benchmark: AnyBenchmark, suite: BenchmarkSuite, settings: BenchmarkSettings
+    ) -> [Double] {
+        var measurements: [Double] = []
+        var n: Int = 1
+
+        while true {
+            measurements = doNIterations(n, benchmark: benchmark, suite: suite)
```

Comment:
It looks like this algorithm throws away past runs... shouldn't it just keep appending to the existing measurements until you've collected enough instead?

Reply:
From looking through the reference implementation, I believe this is consistent with google/benchmark. We only want to record results of directly contiguous iterations to avoid any potential interference when going in and out of the tight loop.

Reply:
This might increase warmup iterations above the minimum ones, but it's usually a good thing, given that our primary goal is to evaluate steady state performance.